draco/fact_utils.py

```diff
@@ -78,7 +80,134 @@ def dict_to_facts(
                 yield make_fact(FactKind.ATTRIBUTE, (path, parent, data))
 
 
+def deep_nest(Root: Dict[Union[str, int], Any], current_dict: Dict):
+    """
+    Indexes into the memo dictionary and creates the nested data structure
+    """
+    for key in current_dict:
+        if isinstance(current_dict[key], list):
+            current_list = current_dict[key]
+            for i in range(len(current_list)):
+                if isinstance(current_list[i], int) and current_list[i] in Root:
+                    index = current_list[i]
+                    current_list[i] = Root[index]
+                deep_nest(Root, current_list[i])
+        elif isinstance(current_dict[key], Mapping):
+            deep_nest(Root, current_dict[key])
+    return
+
+
+def remove_memo(Root: Dict[Union[str, int], Any]):
+    """
+    Removes the integer ID mappings in the dictionary
+    """
+    to_delete = []
+    for key in Root:
+        if isinstance(key, int):
+            to_delete.append(key)
+
+    for key in to_delete:
+        del Root[key]
+
+    return
+
+
+def parse_values(values: List) -> List:
+    """
+    If the value is a number, the function converts the string to a nuumber
+    and return the string otherwise
+    """
+    result = []
+    for value in values:
+        if value.type == SymbolType.Function:
+            name, args = value.name, value.arguments
+            if args == []:
+                result.append(name)
+            else:
+                result.append(tuple(parse_values(list(args))))
+        else:
+            result.append(value.number)
+    return result
+
+
+def handle_path(
+    address: List[Union[str, int]], value: Any, nested_dict: Dict, kind: FactKind
+) -> Dict[Union[str, int], Any]:
+    assert len(address) != 0
+    if len(address) == 1:
+        last = address[0]
+        if kind == FactKind.ATTRIBUTE.value:
+            nested_dict[last] = value
+        else:
+            if last in nested_dict:
+                nested_dict[last].append(value)
+            else:
+                nested_dict[last] = [value]
+        return nested_dict
+
+    else:
+        elem = address[0]
+        if elem in nested_dict:
+            nested_dict[elem] = handle_path(address[1:], value, nested_dict[elem], kind)
+        else:
+            nested_dict[elem] = handle_path(address[1:], value, dict(), kind)
+        return nested_dict
+
+
 def facts_to_dict(facts: List) -> Mapping:
-    """A generic decoder that converts an answer set into a nested data structure"""
-    pass  # TODO: https://github.com/cmudig/draco2/issues/24
-    return {}
+    """
+    A generic decoder that converts an answer set into a nested data structure
+
+    """
```

Comment:
This code needs more explanations or be a bit simplified. I am having a hard time reading it.